### PR TITLE
TGroup library to manage the priority of xapi execution threads.

### PR DESCRIPTION
**Pull request by GabrielBuica**
Currently xapi threads are running with the same priority causing slowness when a lot of threads are waiting for locks. 

I've created this draft to discuss the classification of groups of different xapi execution threads. The initial plan is to classify the server threads, smapi requests, and external requests while having the default running as normal.

The `Group` module is the general abstraction for xapi threads, and the `Cgroup` module is the specialized in the management of groups as control groups. In the end it will probably have separate modules for managing the priority and nice levels.

### CODE REVIEW

File: ocaml/libs/tgroup/tgroup.ml [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+ * but WITHOUT Group WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+module D = Debug.Make (struct let name = __MODULE__ end)
+
+let ( // ) = Filename.concat
+
+let requests = "/sys/fs/cgroup/cpu/control.slice/xapi.service/request"
```
Comment (psafont) [inline ocaml/libs/tgroup/tgroup.ml:21]: I don't think this value should be hardcoded, the interface doesn't make it seem like it should just work for the xapi service.

File: ocaml/libs/tgroup/tgroup.ml [hunk 2]
```diff
@@ -119,4 +134,37 @@ module Cgroup = struct
     Group.all
     |> List.map dir_of
     |> List.iter (fun dir -> Xapi_stdext_unix.Unixext.mkdir_rec dir 0o755)
+
+  let write_line_to_file filename s =
```
Comment (psafont) [inline ocaml/libs/tgroup/tgroup.ml:138]: Could you use any of the uniext function to implement this? https://github.com/xapi-project/xen-api/blob/master/ocaml/libs/xapi-stdext/lib/xapi-stdext-unix/unixext.mli#L83

File: ocaml/libs/tgroup/tgroup.ml (hunk 1 above)
Comment (GabrielBuica) [reply ocaml/libs/tgroup/tgroup.ml:21]: I was thinking of adding it in the unixext. Or a version of it. It's basically `pidfile_write` but the pid (in this case I want to pass the tid as a parameter) is passed as a parameter. I will also need something similar to write the cpu shares for cgroups.

File: ocaml/libs/tgroup/tgroup.ml (hunk 2 above)
Comment (GabrielBuica) [reply ocaml/libs/tgroup/tgroup.ml:138]: I was thinking of adding it in the unixext. Or a version of it. It's basically pidfile_write but the pid (in this case I want to pass the tid as a parameter) is passed as a parameter. I will also need something similar to write the cpu shares for cgroups.

File: ocaml/libs/tgroup/tgroup.mli [hunk 3]
```diff
@@ -46,12 +46,23 @@ module Cgroup : sig
...
+  originator [o].*)
 end
+
+val of_originator : Group.Originator.t -> unit
+(** [of_originator o] classifies the crrent thread accordingly to [Group]
+classification and *)
+
+val of_req_originator : string option -> unit
+(** [of_req_originator o] same as [of_originator] but it does base on the http 
+request header. *)
```
Comment (last-genius) [inline ocaml/libs/tgroup/tgroup.mli:68]: seems like this reformatting should be in the first commit

File: ocaml/libs/tgroup/tgroup.mli [hunk 4]
```diff
@@ -46,12 +46,23 @@ module Cgroup : sig
...
   val init : unit -> unit
-  (** [init ()] initializes the hierachy of cgroups associated to all [Group.t] types.
+  (** [init ()] initializes the hierachy of cgroups associated to all [Group.t]
+  types.
   *)
 
   val set_cgroup : Group.Originator.t -> unit
-  (** [set_cgroup o] sets the current xapi thread in a cgroup based on the originator [o].*)
+  (** [set_cgroup o] sets the current xapi thread in a cgroup based on the
+  originator [o].*)
```
Comment (last-genius) [inline ocaml/libs/tgroup/tgroup.mli:59]: seems like this reformatting should be in the first commit

Comment (GabrielBuica) [reply ocaml/libs/tgroup/tgroup.mli]: I modified that, and some spelling and grammar.

File: ocaml/libs/tgroup/tgroup.ml (hunk 1 above)
Comment (GabrielBuica) [reply ocaml/libs/tgroup/tgroup.ml]: I changed `init` to accept a directory and plan to add this in `Xapi_globs`

File: ocaml/libs/tgroup/tgroup.ml [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+  let write_cur_tid_to_cgroup_file filename =
+    let fd =
+      Unix.openfile filename [Unix.O_WRONLY; Unix.O_CREAT; Unix.O_TRUNC] 0o640
+    in
+    Xapi_stdext_pervasives.Pervasiveext.finally
+      (fun () ->
+        let buf = "0\n" in
+        let len = String.length buf in
+        if Unix.write fd (Bytes.unsafe_of_string buf) 0 len <> len then
+          debug "writing current tid to %s failed" filename
```
Comment (lindig) [inline ocaml/libs/tgroup/tgroup.ml:149]: This should be a warning or error

File: ocaml/libs/tgroup/tgroup.mli [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+  (** Abstract type that represents a group of execution threads in xapi. Each
+  group corresponds to a Creator, and has a designated level of priority.*)
+  type t
+
+  (** Generic representation of different xapi threads originators. *)
+  module Originator : sig
+    (** Type that represents different originators of xapi threads. *)
+    type t = Internal_Host_SM | EXTERNAL | Internal_Server
+
+    val of_string : string -> t
```
Comment (lindig) [inline ocaml/libs/tgroup/tgroup.mli:27]: Whats strings are expected here?

**Review by lindig — COMMENTED** [review]
Is GADT magic really required?

File: ocaml/libs/tgroup/tgroup.mli (hunk 6 above)
Comment (GabrielBuica) [reply ocaml/libs/tgroup/tgroup.mli:27]: My intention is to have a new http header, `originator`, based on which I set the priority of requests. Initially `SM` would send this header and (the `CLI` would be added next). So this function would be called on the values of that specific header.

Not having a mapping to `Internal_server` was intentional, as I didn't want the possibility of an http request to mimic a Server priority.  Classifying the server threads would use `Originator.Internal_Server` directly.

**Comment by GabrielBuica** [comment]
As for why GADT? The classification is not yet set in stone and most likely will have some changes. I wanted to have exhaustive pattern matching without using the wildcard so when a new group is added, or a change is being made, all the necessary changes will be highlighted by dune as non-exhaustive.

Right now, the `Creator` type has only one field, `originator`, but the classification will most likely end by also taking into account types of users and endpoints as well. All these field would have a finite set of values that we are interested in. 

Initially I was using strings for all of them but it required quite a lot of attention to make the changes everywhere correctly. Afterwards, I tried to define variants, which helped, but this time I had a lot of variants defined everywhere that would depend on each other. Ended up with a similar problem, having a lot of variant types and every time I would tweak something I had to add a new variant tag in a lot of places. Finally, I ended up with the current `Group` module that uses gadts, I tried to limit their use (and hide it from outside of the library implementation) as much as possible.

**Comment by GabrielBuica** [comment]
Closing this as I'll make a PR for feature/perf.

**Pull request closed without merging**